pubmed-semantic(updated).py

```diff
@@ -0,0 +1,286 @@
+        id = uids[lower:upper]
+        try:
+            eutil_response = _get_eutil_records("efetch", id)
+            ERROR = _get(eutil_response, ["eFetchResult", "ERROR"])
+            if ERROR:
+                raise RuntimeError(ERROR)
+            pubmedEfetchReponse = PubmedEfetchResponse(**eutil_response)
+            pubmedArticle = pubmedEfetchReponse.PubmedArticleSet.PubmedArticle
+        except Exception as e:
+            print(f"Error encountered in uids_to_docs {e}")
+            raise e
+        else:
+            articles = pubmedArticle if isinstance(pubmedArticle, list) else [pubmedArticle]
+            output = _articles_to_docs(articles)
+            docs = docs + output
+    return docs
+
+def get_list_pmid(start, end):
+    #using the esearch eutil to grab the 'Webenv' and 'total record' values to then send to efetch
+    search_url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&mindate="+start+"&maxdate="+end+"&usehistory=y&retmode=json"
+    search_r = requests.post(search_url)
+    data = search_r.json()
+    webenv = data["esearchresult"]['webenv']
+    total = int(data["esearchresult"]['count'])
+    #using the efetch eutil with the 'Webenv' from esearch to grab all the data
+    fetch_url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmax=9999&query_key=1&WebEnv="+webenv
+    #Total records found in requested time frame
+    print("Total records :"+str(total))
+    #count used for URL number
+    count = 1
+    store = []
+    #First loop to grab all the data and write it to .json files, each file only has 10,000 records
+    for i in range(0, total, 10000):
+        #changes the retstart value by 10,000 every iteration so we don't print the same records over again
+        this_fetch = fetch_url+"&retstart="+str(i)
+        #URL that can be used in browser for data viewing
+        print("URL "+str(count)+": "+this_fetch)
+        count += 1
+        fetch_r = requests.post(this_fetch)
+        #Writing all the data to .json files (1 file per 10,000 records)
+        f = open('pubmed_batch_'+str(i)+'_to_'+str(i+9999)+".json", 'w')
+        f.write(fetch_r.text)
+        f.close()
+    for i in range(0, total, 10000):
+        f = open('pubmed_batch_'+str(i)+'_to_'+str(i+9999)+".json", 'r')
+        data_file = f.read()
+        p_flag = False
+        for word in data_file.split():
+            if word == 'pmid':
+                p_flag = True
+            elif p_flag == True:
+                p_flag = False
+                if word[:-1] not in store:
+                    store.append(word[:-1])
+        f.close()
+    return store
+
+def main():
+    ab = open("test.json", 'w')
+    store = get_list_pmid("2021/01/17","2021/01/17")
+    print(store)
+    print(len(store))
+    here = uids_to_docs(store)
+    for item in here:
+        ab.write(json.dumps(item))
+        ab.write("\n")
+    ab.close()
+
+if __name__ == "__main__":
+    main()
```

Comment:
Okay so in general we should keep the code and the tests in different files (we can talk about setting up unit tests sometime if you like!).

I would remove this. If you like, you can create a docstring for `get_list_pmid` and put a simple hardcoded example in there.

Reply:
Okay I see, I can use another file to test it. Sure, we can talk about unit tests when you are free, I appreciate that.

Reply:
The short version: I have a strong preference for [`pytest`](https://docs.pytest.org/en/stable/). It is very easy to pick up and get started with.